File: CS/Basic/recursion/maxBread.go
package main

// import (
// 	"fmt"
// )

func maxBread(money int32, price int32, sticker int32) int32 {
	bread := money / price

	return maxBreadHelper(sticker, bread, bread)
}

func maxBreadHelper(sticker int32, bread int32, holdSticker int32) int32 {

	if sticker > holdSticker {
		return bread
	}

	addBread := holdSticker / sticker
	consumeSticker := sticker * addBread

	return maxBreadHelper(sticker, bread+addBread, holdSticker-consumeSticker+addBread)
}

// func main() {
// 	fmt.Println(maxBread(10, 2, 3))
// }

File: CS/Basic/recursion/MaxBread.java
package CS.Basic.recursion;

public class MaxBread {
    public static int maxBread(int money, int price, int sticker) {
        final int bread = (int) Math.floor(money / price);
        return maxBreadHelper(sticker, bread, bread);
    }

    public static int maxBreadHelper(int sticker, int bread, int hold) {
        if (sticker > hold)
            return bread;
        final int newBread = (int) Math.floor(hold / sticker);
        return maxBreadHelper(sticker, bread + newBread, hold - sticker * newBread + newBread);
    }

    public static void main(String[] args) {
        System.out.println(maxBread(10, 2, 3));
        System.out.println(maxBread(15, 1, 3));
        System.out.println(maxBread(20, 2, 10));
        System.out.println(maxBread(50, 3, 2));
    }
}
